### PR TITLE
[LangRef] Relax semantics of writeonly / memory(write)

### LINKED ISSUE #95152: [MemCpyOpt] Call slot optimization doesn't respect writeonly
```
$ cat /tmp/a.ll
declare void @g(ptr)

define void @f(ptr dead_on_unwind noalias writable writeonly sret(i64) align 8 %0) {
  %a = alloca i64
  call void @g(ptr %a)
  call void @llvm.memcpy(ptr align 8 %0, ptr align 8 %a, i64 8, i1 false)
  ret void
}
$ opt -p memcpyopt -S /tmp/a.ll
declare void @g(ptr)

define void @f(ptr dead_on_unwind noalias writable writeonly sret(i64) align 8 %0) {
  %a = alloca i64, align 8
  call void @g(ptr %0)
  ret void
}
```

If `@g` is not `writeonly` in regards to the `ptr` param, we may end up reading from `%0` in `@g` which is marked `writeonly`.

### PR DESCRIPTION
**Pull request by nikic**
Instead of making writes immediate undefined behavior, consider these attributes in terms of their externally observable effects. We don't care if a location is read within the function, as long as it has no impact on observed behavior. In particular, allow:

 * Reading a location after writing it.
 * Reading a location before writing it (within the function) returns a poison value.

The latter could be further relaxed to also allow things like "reading the value and then writing it back", but I'm not sure how one would specify that operationally (so that proof checkers can verify it).

While here, also explicitly mention the fact that reads and writes to allocas and read from constant globals are `memory(none)`.

Fixes https://github.com/llvm/llvm-project/issues/95152.

### CODE REVIEW

File: llvm/docs/LangRef.rst [hunk 1]
```diff
@@ -1973,6 +1975,21 @@ example:
     - ``memory(readwrite, argmem: none)``: May access any memory apart from
       argument memory.
 
+    The supported access kinds are:
+
+    - ``readwrite``: Any kind of access to the location is allowed.
+    - ``read``: The location is only read. Writing the location is immediate
```
Comment (aeubanks) [inline llvm/docs/LangRef.rst:1981]: "writing the location" -> "writing to the location"

File: llvm/docs/LangRef.rst [hunk 2]
```diff
@@ -1973,6 +1975,21 @@ example:
...
+      undefined behavior. This includes the case where the location is read and
```
Comment (aeubanks) [inline llvm/docs/LangRef.rst:1982]: "location is read" -> "location is read from"

**Review by aeubanks — APPROVED** [review]
thanks, this makes sense

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 3]
```diff
@@ -2359,20 +2359,20 @@ static Instruction *foldSelectCmpBitcasts(SelectInst &Sel,
 /// operand, the result of the select will always be equal to its false value.
 /// For example:
 ///
-///   %0 = cmpxchg i64* %ptr, i64 %compare, i64 %new_value seq_cst seq_cst
-///   %1 = extractvalue { i64, i1 } %0, 1
-///   %2 = extractvalue { i64, i1 } %0, 0
-///   %3 = select i1 %1, i64 %compare, i64 %2
-///   ret i64 %3
+///   %cmpxchg = cmpxchg ptr %ptr, i64 %compare, i64 %new_value seq_cst seq_cst
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:2362]: I don't think this commit is related to the PR.

Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp]: Yeah, that was not supposed to be there... Dropped the commit!

**Pull request merged**